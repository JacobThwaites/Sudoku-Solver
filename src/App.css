.alert {
    width: 400px;
    justify-content: center;
    margin-left: 70px;
}

body {
    background-color: #F9F9F9;
}