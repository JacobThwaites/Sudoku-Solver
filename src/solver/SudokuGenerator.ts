import { SudokuGridType } from "../SudokuGridType";
import { getEmptyRowsWithNulls } from "../utils/emptyRowGenerator";
import shuffle from "../utils/utils";

type Coordinate = [number, number];

export class SudokuGenerator {
  board: SudokuGridType;

  constructor() {
    this.board = getEmptyRowsWithNulls();
  }

  public generateRandomPuzzle(): SudokuGridType | boolean {
    const randomSequence = this.generateRandomSequence();
    let solution = this.solveUsingSequence(randomSequence);

    if (typeof solution === 'boolean') {
        return false;
    }
    
    const puzzle = this.removeNumbers(solution);

    return puzzle;
  }

  private generateRandomSequence(): number[] {
    const sequence = [1, 2, 3, 4, 5, 6, 7, 8, 9];
    return shuffle(sequence);
  }

  private solveUsingSequence(sequence: number[]): SudokuGridType | boolean {
    const emptySquare: any = this.findEmptySquare();

    if (!emptySquare) {
      return this.board;
    }

    const row = emptySquare[0];
    const col = emptySquare[1];

    for (let i = 0; i < sequence.length; i++) {
      if (this.canAddNum(sequence[i], row, col)) {
        this.board[row][col] = sequence[i];

        if (this.solveUsingSequence(sequence)) {
          return this.board;
        }

        this.board[row][col] = null;
      }

      sequence = this.generateRandomSequence();
    }

    return false;
  }

  private findEmptySquare(): [number, number] | boolean {
    for (let i = 0; i < this.board.length; i++) {
      const row = this.board[i];
      for (let j = 0; j < row.length; j++) {
        if (row[j] === null) {
          return [i, j];
        }
      }
    }
    return false;
  }

  private canAddNum(num: number, row: number, col: number): boolean {
    return (
      !this.isNumInRow(num, row) &&
      !this.isNumInCol(num, col) &&
      !this.isNumInSquare(num, row - (row % 3), col - (col % 3))
    );
  }

  private isNumInRow(num: number, row: number): boolean {
    for (let i = 0; i < 9; i++) {
      if (this.board[row][i] === num) {
        return true;
      }
    }

    return false;
  }

  private isNumInCol(num: number, col: number): boolean {
    for (let i = 0; i < 9; i++) {
      if (this.board[i][col] === num) {
        return true;
      }
    }

    return false;
  }

  private isNumInSquare(num: number, row: number, col: number): boolean {
    for (let i = 0; i < 3; i++) {
      for (let j = 0; j < 3; j++) {
        if (this.board[i + row][j + col] === num) {
          return true;
        }
      }
    }

    return false;
  }

  private removeNumbers(board: SudokuGridType): SudokuGridType {
    const puzzle = board;
    const TOTAL_NUMS_TO_REMOVE = 60;
    const coordinates = this.getShuffledCoordinates()

    for (let i = 0; i < TOTAL_NUMS_TO_REMOVE; i++) {
        const coordinate = coordinates.pop();
        puzzle[coordinate[0]][coordinate[1]] = null;
    }
    
    return puzzle;
  }

  private getShuffledCoordinates(): Coordinate[] {
    const coordinates = [];

    for (let i = 0; i < this.board.length; i++) {
        for (let j = 0; j < this.board[i].length; j++) {
            const coordinate: Coordinate = [i,j];
            coordinates.push(coordinate);
        }
    }
    
    return shuffle(coordinates);
  }
}
