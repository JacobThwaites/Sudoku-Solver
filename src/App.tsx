import { useRef, useState, useEffect } from "react";
import Sudoku from "./sudoku/Sudoku";
import { emptyGrid } from "./utils/emptyRowGenerator";
import { SudokuGridType } from "./SudokuGridType";
import SudokuButtons from "./sudoku/SudokuButtons";
import Alert from '@mui/material/Alert';
import './App.css';
import TopBar from "./TopBar";
import { calculateDelayTime } from "./utils/utils";
import { SudokuGenerator } from "./solver/SudokuGenerator";

export default function App(): JSX.Element {
  const [rows, setRows] = useState(emptyGrid);
  const [solutionSteps, setSolutionSteps] = useState([]);
  const [startingPosition, setStartingPosition] = useState(emptyGrid);
  const [shouldDisplaySolution, setShouldDisplaySolution] = useState(false);
  const displayingSolution = useRef(false);
  const [isSolveButtonPressed, setIsSolveButtonPressed] = useState(false);
  const [noSolutionFound, setNoSolutionFound] = useState(false);
  const [isInvalidPuzzle, setIsInvalidPuzzle] = useState(false);
  const [speed, setSpeed] = useState('Normal');
  const currentSpeed = useRef('Normal');
  
  // Allows interrupting the for loop for displaying the solution if clear button is pressed.
  useEffect(() => {
    displayingSolution.current = shouldDisplaySolution;
  }, [shouldDisplaySolution]);

  useEffect(() => {
    currentSpeed.current = speed;
  }, [speed]);

  function setSquareValue(num: number, row: number, col: number): void {
    const newRows = [...rows];
    newRows[row][col] = num;
    setRows(newRows);
  };

  function updateStartingPosition(startingPosition: SudokuGridType): void {
    setStartingPosition(startingPosition);
  }

  function onSolveButtonPressed(rows: SudokuGridType): void {
    setIsSolveButtonPressed(true);
    setIsInvalidPuzzle(false);
    setRows(rows);
  }

  function onClearButtonPressed(): void {
    clearRows();
    setIsSolveButtonPressed(false);
    setNoSolutionFound(false);
    setIsInvalidPuzzle(false);
  }

  function clearRows(): void {
    const newRows = [...rows];

    for (let i = 0; i < newRows.length; i++) {
      for (let j = 0; j < newRows[i].length; j++) {
        newRows[i][j] = null;
      }
    }
    
    setShouldDisplaySolution(false);
    setRows(newRows);
    setStartingPosition(newRows);
    setSolutionSteps([]);
  }

  async function showSolutionSteps() {
    setShouldDisplaySolution(true);
    setRowsToStartingPosition();

    for (let i = 0; i < solutionSteps.length; i++) {
      const delayTime = calculateDelayTime(currentSpeed.current);
      await delay(delayTime);
      
      if (!displayingSolution.current) {
        break;
      }

      const step = solutionSteps[i];
      addStepToGrid(step);
    }
  }

  function setRowsToStartingPosition(): void {
    const newRows = [...rows];

    for (let i = 0; i < newRows.length; i++) {
      for (let j = 0; j < newRows[i].length; j++) {
        newRows[i][j] = startingPosition[i][j];
      }
    }

    setRows(newRows);
  }

  function delay(ms: number): Promise<number> {
    return new Promise(res => setTimeout(res, ms));
  } 

  function addStepToGrid(step: number[]): void {
    const row = step[0];
    const col = step[1];
    const num = step[2];
    setSquareValue(num, row, col);
  }

  function generateRandomPuzzle(): void {
    onClearButtonPressed();

    const generator = new SudokuGenerator();
    const puzzle = generator.generateRandomPuzzle();

    if (typeof puzzle !== 'boolean') {
      setPuzzle(puzzle);
    }
  }

  function setPuzzle(puzzle: SudokuGridType): void {
    for (let i = 0; i < puzzle.length; i++) {
      for (let j = 0; j < puzzle[i].length; j++) {
        setSquareValue(puzzle[i][j], i, j); 
      }        
    }
  }

  return (
    <div>
      <TopBar speed={speed} setSpeed={setSpeed} onGenerateButtonPressed={generateRandomPuzzle}/>
      <Sudoku 
        setSquareValue={setSquareValue} 
        rows={rows} 
        startingPosition={startingPosition}
      />
      {isInvalidPuzzle &&
        <Alert className='alert' severity="error">Invalid puzzle!</Alert>
      }
      {noSolutionFound &&
        <Alert severity="error">No valid solution found!</Alert>
      }
      <SudokuButtons 
        isSolveButtonDisabled={isSolveButtonPressed}
        rows={rows}
        updateStartingPosition={updateStartingPosition}
        onSolveButtonPressed={onSolveButtonPressed}
        setSolutionSteps={setSolutionSteps}
        onClearButtonPressed={onClearButtonPressed}
        isShowStepsButtonDisabled={!solutionSteps.length}
        onSolutionButtonClick={showSolutionSteps}
        setNoSolutionFound={setNoSolutionFound}
        setIsInvalidPuzzle={setIsInvalidPuzzle}
      />
    </div>
  );
}