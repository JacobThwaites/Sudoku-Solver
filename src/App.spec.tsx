import { render, fireEvent, getByTestId, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import App from './App';
import { SudokuGenerator } from './solver/SudokuGenerator';


it('renders App correctly', () => {
    const component = renderer.create(
        <App />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

it('renders alert if puzzle is incorrect', () => {
    const { getAllByTestId, getByText } = render(<App />);
    const squares = getAllByTestId('square')
    for (let i = 0; i < squares.length; i++) {
        fireEvent.keyDown(squares[i], { key: '3', keyCode: 51 })
    }

    const solveButton = getByText('Solve');
    fireEvent.click(solveButton);

    const invalidPuzzleAlert = getByText('Invalid puzzle!');
    expect(invalidPuzzleAlert).toBeInTheDocument();
});

it('clears puzzle on clear button press', () => {
    const { getAllByTestId, getByText } = render(<App />);
    const squares = getAllByTestId('square')
    for (let i = 0; i < squares.length; i++) {
        fireEvent.keyDown(squares[i], { key: '3', keyCode: 51 })
    }

    for (let i = 0; i < squares.length; i++) {
        expect(squares[i]).toHaveValue('3')
    }

    const clearButton = getByText('Clear');
    fireEvent.click(clearButton);

    for (let i = 0; i < squares.length; i++) {
        expect(squares[i]).toHaveValue('')
    }
});

it('only enables show steps button if puzzle is solved', () => {
    const { getByText } = render(<App />);

    const showStepsButton = getByText('Show Steps');
    expect(showStepsButton).toBeDisabled();

    const solveButton = getByText('Solve');
    fireEvent.click(solveButton);

    expect(showStepsButton).toBeEnabled();
});

it('can generate a random puzzle', () => {
    jest.spyOn(SudokuGenerator.prototype, 'generateRandomPuzzle').mockImplementation(() => false);
    const { getByText } = render(<App />);

    const generateButton = getByText('Generate Random Puzzle');
    fireEvent.click(generateButton);

    expect(SudokuGenerator.prototype.generateRandomPuzzle).toHaveBeenCalledTimes(1);
});