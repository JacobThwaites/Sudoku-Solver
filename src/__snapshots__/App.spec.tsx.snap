// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders App correctly 1`] = `
<div>
  <header
    className="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionSticky css-7e0s21-MuiPaper-root-MuiAppBar-root"
    id="top-bar"
  >
    <svg
      id="logo"
    >
      sudoku-logo.svg
    </svg>
    <div
      className="MuiTypography-root MuiTypography-h6 css-15zqwrc-MuiTypography-root"
    >
      Sudoku Solver
    </div>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
      disabled={false}
      id="generate-puzzle-button"
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      Generate Random Puzzle
    </button>
    <div
      className="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink={true}
        id="speed-select-label"
      >
        Speed:
      </label>
      <div
        className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
        data-testid="speed-select"
        onClick={[Function]}
      >
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-labelledby="speed-select-label speed-select"
          className="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
          id="speed-select"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          role="button"
          tabIndex={0}
        >
          Normal
        </div>
        <input
          aria-hidden={true}
          className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
          disabled={false}
          onAnimationStart={[Function]}
          onChange={[Function]}
          required={false}
          tabIndex={-1}
          value="Normal"
        />
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
          data-testid="ArrowDropDownIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
        <fieldset
          aria-hidden={true}
          className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="css-1in441m"
          >
            <span>
              Speed
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </header>
  <div
    id="sudoku"
  >
    <div
      className="row"
    >
      <input
        className="square top-square left-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square top-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square top-square right-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square top-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square top-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square top-square right-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square top-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square top-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square top-square right-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
    </div>
    <div
      className="row"
    >
      <input
        className="square left-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square right-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square right-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square right-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
    </div>
    <div
      className="row"
    >
      <input
        className="square bottom-square left-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square bottom-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square bottom-square right-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square bottom-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square bottom-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square bottom-square right-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square bottom-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square bottom-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square bottom-square right-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
    </div>
    <div
      className="row"
    >
      <input
        className="square left-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square right-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square right-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square right-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
    </div>
    <div
      className="row"
    >
      <input
        className="square left-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square right-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square right-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square right-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
    </div>
    <div
      className="row"
    >
      <input
        className="square bottom-square left-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square bottom-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square bottom-square right-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square bottom-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square bottom-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square bottom-square right-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square bottom-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square bottom-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square bottom-square right-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
    </div>
    <div
      className="row"
    >
      <input
        className="square left-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square right-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square right-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square right-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
    </div>
    <div
      className="row"
    >
      <input
        className="square left-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square right-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square right-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square right-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
    </div>
    <div
      className="row"
    >
      <input
        className="square bottom-square left-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square bottom-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square bottom-square right-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square bottom-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square bottom-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square bottom-square right-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square bottom-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square bottom-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
      <input
        className="square bottom-square right-square"
        data-testid="square"
        maxLength={1}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    id="sudoku-buttons"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium styled-button css-sghohy-MuiButtonBase-root-MuiButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      Solve
    </button>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium styled-button css-sghohy-MuiButtonBase-root-MuiButton-root"
      disabled={false}
      id="clear-button"
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      Clear
    </button>
    <button
      className="MuiButtonBase-root Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium styled-button css-sghohy-MuiButtonBase-root-MuiButton-root"
      disabled={true}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={-1}
      type="button"
    >
      Show Steps
    </button>
  </div>
</div>
`;
